### PR TITLE
Client certs

### PR DESCRIPTION
**Pull request by DieMyst**

### CODE REVIEW

File: client/fluence-js/src/fluence_connection.ts [hunk 1]
```diff
@@ -162,7 +162,7 @@ export class FluenceConnection {
 
     private async sendCall(call: FunctionCall) {
         let callStr = callToString(call);
-        console.log("send function call: " + callStr);
+        console.log("send function call: " + JSON.stringify(JSON.parse(callStr), undefined, 2));
```
Comment (folex) [inline client/fluence-js/src/fluence_connection.ts:165]: Why do this? `JSON.stringify(JSON.parse(callStr), undefined, 2)`

File: client/fluence-js/src/test/address.spec.ts [hunk 2]
```diff
@@ -89,14 +92,69 @@ describe("Typescript usage suite", () => {
...
+1589974723504`
+
+        let deser = await certificateFromString(cert);
+        let ser = certificateToString(deser);
+
+        expect(ser).to.be.equal(cert);
     });
+
+    // uncomment it and run `npm run test` to check service's and certificate's api with Fluence nodes
+    /*it("integration test", async function () {
```
Comment (folex) [inline client/fluence-js/src/test/address.spec.ts:114]: `it.skip`?

File: client/fluence-js/src/test/address.spec.ts [hunk 3]
```diff
@@ -89,14 +92,69 @@ describe("Typescript usage suite", () => {
...
 
+export async function testCerts() {
+    let key1 = await Fluence.generatePeerId();
+    let key2 = await Fluence.generatePeerId();
+
+    // connect to two different nodes
+    let cl1 = await Fluence.connect("/dns4/104.248.25.59/tcp/9003/ws/p2p/12D3KooWBUJifCTgaxAUrcM9JysqCcS4CS8tiYH5hExbdWCAoNwb", key1);
+    let cl2 = await Fluence.connect("/ip4/104.248.25.59/tcp/9002/ws/p2p/12D3KooWHk9BjDQBUqnavciRPhAYFvqKBe4ZiPPvde7vDaqgn5er", key2);
+
+    let certGiver1 = new TrustGraph(cl1);
```
Comment (folex) [inline client/fluence-js/src/test/address.spec.ts:131]: Why `Giver`? `certPublisher1` / `trustGraph1`

File: client/fluence-js/src/trust/certificate.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    if (!lines[lines.length - 1]) {
+        lines.pop()
+    }
+
+    // TODO do match different formats and versions
+    let _format = lines[0];
+    let _version = lines[1];
+    console.log("LENGTH: " + lines.length)
+
+    if ((lines.length - 2) % 4 !== 0) {
```
Comment (folex) [inline client/fluence-js/src/trust/certificate.ts:46]: Why `- 2 % 4`? Add comment

File: client/fluence-js/src/trust/certificate.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+                            extendCert: Certificate,
+                            expiresAt: number,
+                            issuedAt: number): Promise<Certificate> {
+    if (expiresAt < issuedAt) {
+        throw Error("Expiration time should be greater then issued time.")
+    }
+
+    let lastTrust = extendCert.chain[extendCert.chain.length - 1];
+
+
```
Comment (folex) [inline client/fluence-js/src/trust/certificate.ts:94]: Why `\n\n`?

File: client/fluence-js/src/trust/certificate.ts [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    if (lastTrust.issuedFor !== issuedBy) {
+        throw Error("Last trust in chain should be same as 'issuedBy'.")
```
Comment (folex) [inline client/fluence-js/src/trust/certificate.ts:96]: I think error message is a bit hard to understand. `issuedFor should be equal to issuedBy in the last trust of the chain`

File: client/fluence-js/src/trust/certificate.ts [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    let trust = await createTrust(forPk, issuedBy, expiresAt, issuedAt);
+
+    let chain = [...extendCert.chain];
+    chain.push(trust);
```
Comment (folex) [inline client/fluence-js/src/trust/certificate.ts:102]:
```suggestion
    let chain = [...extendCert.chain];
    let trust = await createTrust(forPk, issuedBy, expiresAt, issuedAt);
    chain.push(trust);
```

So many empty lines...........

File: client/fluence-js/src/trust/trust.ts [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        issuedFor: peerId,
+        signature: signature,
+        expiresAt: parseInt(expiresAt),
+        issuedAt: parseInt(issuedAt)
+    }
+}
+
+export async function createTrust(forPk: PeerId, issuedBy: PeerId, expiresAt: number, issuedAt: number): Promise<Trust> {
+    let bytes = toSignMessage(forPk, expiresAt, issuedAt);
+
```
Comment (folex) [inline client/fluence-js/src/trust/trust.ts:48]: why \n

File: client/fluence-js/src/trust/trust.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    let signature = await issuedBy.privKey.sign(Buffer.from(bytes));
+
```
Comment (folex) [inline client/fluence-js/src/trust/trust.ts:50]: why \n

File: client/fluence-js/src/trust/trust_graph.ts [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    // Publish certificate to Fluence network. It will be published in Kademlia neighbourhood by `peerId` key.
+    async publishCertificates(peerId: string, certs: Certificate[]) {
+        let certsStr = [];
+        for (let cert of certs) {
+            certsStr.push(await certificateToString(cert));
+        }
+
+        await this.client.sendServiceCall("add_certificates", {
```
Comment (folex) [inline client/fluence-js/src/trust/trust_graph.ts:37]: I think it's better to wait for a reply with success or error

File: client/fluence-js/src/test/address.spec.ts [hunk 11]
```diff
@@ -89,14 +92,69 @@ describe("Typescript usage suite", () => {
...
+    expiresAt.setDate(new Date().getDate() + 1);
+
+    // create root certificate for key1 and extend it with key2
+    let rootCert = await nodeRootCert(key1);
+    let extended = await issue(key1, key2, rootCert, expiresAt.getTime(), issuedAt.getTime());
+
+    // publish certificates to Fluence network
+    await certGiver1.publishCertificates(key2.toB58String(), [extended]);
+
+    await delay(2000);
```
Comment (folex) [inline client/fluence-js/src/test/address.spec.ts:146]: No need to wait for 2 seconds if you would wait for reply in `publishCertificates`

File: client/fluence-js/src/trust/certificate.ts [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+                                forPk: PeerId,
+                                expiresAt: number,
+                                issuedAt: number,
+): Promise<Certificate> {
+    if (expiresAt < issuedAt) {
+        throw Error("Expiration time should be greater then issued time.")
+    }
+
+    let maxDate = new Date(158981172690500).getTime();
+
```
Comment (folex) [inline client/fluence-js/src/trust/certificate.ts:71]: why \n

File: client/fluence-js/src/trust/certificate.ts [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+    let rootTrust = await createTrust(issuedBy, issuedBy, maxDate, issuedAt);
+    let trust = await createTrust(forPk, issuedBy, expiresAt, issuedAt);
+
```
Comment (folex) [inline client/fluence-js/src/trust/certificate.ts:74]: why \n

File: client/fluence-js/src/fluence_connection.ts (hunk 1 above)
Comment (DieMyst) [reply client/fluence-js/src/fluence_connection.ts:165]: to log in formatted json

**Pull request merged**